File: genai_telemetry/core/decorators.py
"""
Decorators for tracing LLM operations.
"""

import time
from functools import wraps
from typing import Callable

from genai_telemetry.core.utils import (
    extract_tokens_from_response,
    extract_content_from_response,
)


def _get_telemetry():
    """Import telemetry to avoid circular imports."""
    from genai_telemetry.core.telemetry import get_telemetry
    return get_telemetry()


def trace_llm(
    model_name: str,
    model_provider: str = "openai",
    extract_content: bool = False
) -> Callable:
    """
    Decorator to trace LLM calls.
    
    IMPORTANT: To capture token usage, you must return the FULL response object
    from the LLM client, not just response.choices[0].message.content.
    
    Args:
        model_name: Name of the model (e.g., "gpt-4o", "claude-3-opus")
        model_provider: Provider name (e.g., "openai", "anthropic")
        extract_content: If True, automatically extract and return just the content
                        string while still capturing token metrics. This allows you
                        to return the full response in your function but get just
                        the content string back.
    
    Example (correct - captures tokens):
        @trace_llm(model_name="gpt-4o", model_provider="openai")
        def chat(message):
            response = client.chat.completions.create(
                model="gpt-4o",
                messages=[{"role": "user", "content": message}]
            )
            return response  # Return FULL response
        
        answer = chat("Hello")
        print(answer.choices[0].message.content)  # Extract content here
    
    Example (with extract_content=True):
        @trace_llm(model_name="gpt-4o", model_provider="openai", extract_content=True)
        def chat(message):
            response = client.chat.completions.create(
                model="gpt-4o",
                messages=[{"role": "user", "content": message}]
            )
            return response  # Return FULL response
        
        answer = chat("Hello")  # answer is now just the content string
        print(answer)
    
    Example (WRONG - tokens will be 0):
        @trace_llm(model_name="gpt-4o", model_provider="openai")
        def chat(message):
            response = client.chat.completions.create(...)
            return response.choices[0].message.content  # DON'T DO THIS
    """
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def wrapper(*args, **kwargs):
            telemetry = _get_telemetry()
            start = time.time()
            
            try:
                result = func(*args, **kwargs)
                duration = round((time.time() - start) * 1000, 2)
                
                # Extract tokens using the helper function
                input_tokens, output_tokens = extract_tokens_from_response(result)
                
                telemetry.send_span(
                    span_type="LLM",
                    name=func.__name__,
                    model_name=model_name,
                    model_provider=model_provider,
                    duration_ms=duration,
                    input_tokens=input_tokens,
                    output_tokens=output_tokens,
                    total_tokens=input_tokens + output_tokens
                )
                
                # If extract_content is True, return just the content string
                if extract_content:
                    return extract_content_from_response(result, model_provider)
                
                return result
                
            except Exception as e:
                duration = round((time.time() - start) * 1000, 2)
                telemetry.send_span(
                    span_type="LLM",
                    name=func.__name__,
                    model_name=model_name,
                    model_provider=model_provider,
                    duration_ms=duration,
                    status="ERROR",
                    is_error=1,
                    error_message=str(e),
                    error_type=type(e).__name__
                )
                raise
        return wrapper
    return decorator


def trace_embedding(model: str) -> Callable:
    """Decorator to trace embedding calls."""
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def wrapper(*args, **kwargs):
            telemetry = _get_telemetry()
            start = time.time()
            
            try:
                result = func(*args, **kwargs)
                duration = round((time.time() - start) * 1000, 2)
                
                # Try to extract token usage from embedding response
                input_tokens = 0
                if hasattr(result, "usage") and result.usage:
                    input_tokens = getattr(result.usage, "prompt_tokens", 0) or 0
                    if not input_tokens:
                        input_tokens = getattr(result.usage, "total_tokens", 0) or 0
                
                telemetry.send_span(
                    span_type="EMBEDDING",
                    name=func.__name__,
                    embedding_model=model,
                    duration_ms=duration,
                    input_tokens=input_tokens
                )
                return result
                
            except Exception as e:
                duration = round((time.time() - start) * 1000, 2)
                telemetry.send_span(
                    span_type="EMBEDDING",
                    name=func.__name__,
                    embedding_model=model,
                    duration_ms=duration,
                    status="ERROR",
                    is_error=1,
                    error_message=str(e)
                )
                raise
        return wrapper
    return decorator


def trace_retrieval(vector_store: str, embedding_model: str = None) -> Callable:
    """Decorator to trace retrieval calls."""
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def wrapper(*args, **kwargs):
            telemetry = _get_telemetry()
            start = time.time()
            
            try:
                result = func(*args, **kwargs)
                duration = round((time.time() - start) * 1000, 2)
                
                docs_count = len(result) if isinstance(result, (list, tuple)) else 0
                
                telemetry.send_span(
                    span_type="RETRIEVER",
                    name=func.__name__,
                    vector_store=vector_store,
                    embedding_model=embedding_model,
                    documents_retrieved=docs_count,
                    duration_ms=duration
                )
                return result
                
            except Exception as e:
                duration = round((time.time() - start) * 1000, 2)
                telemetry.send_span(
                    span_type="RETRIEVER",
                    name=func.__name__,
                    vector_store=vector_store,
                    duration_ms=duration,
                    status="ERROR",
                    is_error=1,
                    error_message=str(e)
                )
                raise
        return wrapper
    return decorator


def trace_tool(tool_name: str) -> Callable:
    """Decorator to trace tool calls."""
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def wrapper(*args, **kwargs):
            telemetry = _get_telemetry()
            start = time.time()
            
            try:
                result = func(*args, **kwargs)
                duration = round((time.time() - start) * 1000, 2)
                
                telemetry.send_span(
                    span_type="TOOL",
                    name=func.__name__,
                    tool_name=tool_name,
                    duration_ms=duration
                )
                return result
                
            except Exception as e:
                duration = round((time.time() - start) * 1000, 2)
                telemetry.send_span(
                    span_type="TOOL",
                    name=func.__name__,
                    tool_name=tool_name,
                    duration_ms=duration,
                    status="ERROR",
                    is_error=1,
                    error_message=str(e)
                )
                raise
        return wrapper
    return decorator


def trace_chain(name: str) -> Callable:
    """Decorator to trace chain/pipeline calls."""
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def wrapper(*args, **kwargs):
            telemetry = _get_telemetry()
            telemetry.new_trace()
            start = time.time()
            
            try:
                result = func(*args, **kwargs)
                duration = round((time.time() - start) * 1000, 2)
                
                telemetry.send_span(
                    span_type="CHAIN",
                    name=name,
                    duration_ms=duration
                )
                return result
                
            except Exception as e:
                duration = round((time.time() - start) * 1000, 2)
                telemetry.send_span(
                    span_type="CHAIN",
                    name=name,
                    duration_ms=duration,
                    status="ERROR",
                    is_error=1,
                    error_message=str(e)
                )
                raise
        return wrapper
    return decorator


def trace_agent(agent_name: str, agent_type: str = None) -> Callable:
    """Decorator to trace agent calls."""
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def wrapper(*args, **kwargs):
            telemetry = _get_telemetry()
            telemetry.new_trace()
            start = time.time()
            
            try:
                result = func(*args, **kwargs)
                duration = round((time.time() - start) * 1000, 2)
                
                telemetry.send_span(
                    span_type="AGENT",
                    name=func.__name__,
                    agent_name=agent_name,
                    agent_type=agent_type,
                    duration_ms=duration
                )
                return result
                
            except Exception as e:
                duration = round((time.time() - start) * 1000, 2)
                telemetry.send_span(
                    span_type="AGENT",
                    name=func.__name__,
                    agent_name=agent_name,
                    agent_type=agent_type,
                    duration_ms=duration,
                    status="ERROR",
                    is_error=1,
                    error_message=str(e)
                )
                raise
        return wrapper
    return decorator

File: js/src/core/decorators.ts
/**
 * Decorators and wrapper functions for tracing LLM operations.
 */

import type { LLMResponse } from '../types';
import { getTelemetry } from './telemetry';
import { extractTokensFromResponse, extractContentFromResponse } from './utils';

/**
 * Wrapper for tracing LLM calls.
 * 
 * @example
 * ```typescript
 * const tracedChat = traceLLM(
 *   async (message: string) => {
 *     return await client.chat.completions.create({
 *       model: 'gpt-4o',
 *       messages: [{ role: 'user', content: message }]
 *     });
 *   },
 *   { modelName: 'gpt-4o', modelProvider: 'openai' }
 * );
 * 
 * const response = await tracedChat('Hello!');
 * ```
 */
export function traceLLM<T extends (...args: unknown[]) => Promise<LLMResponse>>(
  fn: T,
  options: {
    modelName: string;
    modelProvider?: string;
    extractContent?: boolean;
  }
): T {
  const { modelName, modelProvider = 'openai', extractContent = false } = options;

  const wrapped = async (...args: Parameters<T>): Promise<ReturnType<T> | string> => {
    const telemetry = getTelemetry();
    const start = Date.now();

    try {
      const result = await fn(...args);
      const duration = Date.now() - start;

      const [inputTokens, outputTokens] = extractTokensFromResponse(result);

      await telemetry.sendSpan('LLM', fn.name || 'llm_call', {
        modelName,
        modelProvider,
        durationMs: duration,
        inputTokens,
        outputTokens,
      });

      if (extractContent) {
        return extractContentFromResponse(result, modelProvider);
      }

      return result as ReturnType<T>;
    } catch (error) {
      const duration = Date.now() - start;

      await telemetry.sendSpan('LLM', fn.name || 'llm_call', {
        modelName,
        modelProvider,
        durationMs: duration,
        status: 'ERROR',
        isError: 1,
        errorMessage: (error as Error).message,
        errorType: (error as Error).name,
      });

      throw error;
    }
  };

  return wrapped as T;
}

/**
 * Wrapper for tracing embedding calls.
 */
export function traceEmbedding<T extends (...args: unknown[]) => Promise<unknown>>(
  fn: T,
  options: { model: string }
): T {
  const { model } = options;

  const wrapped = async (...args: Parameters<T>): Promise<ReturnType<T>> => {
    const telemetry = getTelemetry();
    const start = Date.now();

    try {
      const result = await fn(...args);
      const duration = Date.now() - start;

      let inputTokens = 0;
      if (result && typeof result === 'object') {
        const resp = result as Record<string, unknown>;
        if (resp.usage && typeof resp.usage === 'object') {
          const usage = resp.usage as Record<string, unknown>;
          inputTokens = (usage.prompt_tokens as number) || (usage.total_tokens as number) || 0;
        }
      }

      await telemetry.sendSpan('EMBEDDING', fn.name || 'embedding', {
        embeddingModel: model,
        durationMs: duration,
        inputTokens,
      });

      return result as ReturnType<T>;
    } catch (error) {
      const duration = Date.now() - start;

      await telemetry.sendSpan('EMBEDDING', fn.name || 'embedding', {
        embeddingModel: model,
        durationMs: duration,
        status: 'ERROR',
        isError: 1,
        errorMessage: (error as Error).message,
      });

      throw error;
    }
  };

  return wrapped as T;
}

/**
 * Wrapper for tracing retrieval calls.
 */
export function traceRetrieval<T extends (...args: unknown[]) => Promise<unknown[]>>(
  fn: T,
  options: { vectorStore: string; embeddingModel?: string }
): T {
  const { vectorStore, embeddingModel } = options;

  const wrapped = async (...args: Parameters<T>): Promise<ReturnType<T>> => {
    const telemetry = getTelemetry();
    const start = Date.now();

    try {
      const result = await fn(...args);
      const duration = Date.now() - start;

      const docsCount = Array.isArray(result) ? result.length : 0;

      await telemetry.sendSpan('RETRIEVER', fn.name || 'retrieval', {
        vectorStore,
        embeddingModel,
        documentsRetrieved: docsCount,
        durationMs: duration,
      });

      return result as ReturnType<T>;
    } catch (error) {
      const duration = Date.now() - start;

      await telemetry.sendSpan('RETRIEVER', fn.name || 'retrieval', {
        vectorStore,
        durationMs: duration,
        status: 'ERROR',
        isError: 1,
        errorMessage: (error as Error).message,
      });

      throw error;
    }
  };

  return wrapped as T;
}

/**
 * Wrapper for tracing tool calls.
 */
export function traceTool<T extends (...args: unknown[]) => Promise<unknown>>(
  fn: T,
  options: { toolName: string }
): T {
  const { toolName } = options;

  const wrapped = async (...args: Parameters<T>): Promise<ReturnType<T>> => {
    const telemetry = getTelemetry();
    const start = Date.now();

    try {
      const result = await fn(...args);
      const duration = Date.now() - start;

      await telemetry.sendSpan('TOOL', fn.name || 'tool_call', {
        toolName,
        durationMs: duration,
      });

      return result as ReturnType<T>;
    } catch (error) {
      const duration = Date.now() - start;

      await telemetry.sendSpan('TOOL', fn.name || 'tool_call', {
        toolName,
        durationMs: duration,
        status: 'ERROR',
        isError: 1,
        errorMessage: (error as Error).message,
      });

      throw error;
    }
  };

  return wrapped as T;
}

/**
 * Wrapper for tracing chain/pipeline calls.
 */
export function traceChain<T extends (...args: unknown[]) => Promise<unknown>>(
  fn: T,
  options: { name: string }
): T {
  const { name } = options;

  const wrapped = async (...args: Parameters<T>): Promise<ReturnType<T>> => {
    const telemetry = getTelemetry();
    telemetry.newTrace();
    const start = Date.now();

    try {
      const result = await fn(...args);
      const duration = Date.now() - start;

      await telemetry.sendSpan('CHAIN', name, {
        durationMs: duration,
      });

      return result as ReturnType<T>;
    } catch (error) {
      const duration = Date.now() - start;

      await telemetry.sendSpan('CHAIN', name, {
        durationMs: duration,
        status: 'ERROR',
        isError: 1,
        errorMessage: (error as Error).message,
      });

      throw error;
    }
  };

  return wrapped as T;
}

/**
 * Wrapper for tracing agent calls.
 */
export function traceAgent<T extends (...args: unknown[]) => Promise<unknown>>(
  fn: T,
  options: { agentName: string; agentType?: string }
): T {
  const { agentName, agentType } = options;

  const wrapped = async (...args: Parameters<T>): Promise<ReturnType<T>> => {
    const telemetry = getTelemetry();
    telemetry.newTrace();
    const start = Date.now();

    try {
      const result = await fn(...args);
      const duration = Date.now() - start;

      await telemetry.sendSpan('AGENT', fn.name || 'agent', {
        agentName,
        agentType,
        durationMs: duration,
      });

      return result as ReturnType<T>;
    } catch (error) {
      const duration = Date.now() - start;

      await telemetry.sendSpan('AGENT', fn.name || 'agent', {
        agentName,
        agentType,
        durationMs: duration,
        status: 'ERROR',
        isError: 1,
        errorMessage: (error as Error).message,
      });

      throw error;
    }
  };

  return wrapped as T;
}

// ============================================================================
// TypeScript Decorators (for class methods)
// ============================================================================

/**
 * Decorator for tracing LLM calls on class methods.
 * 
 * @example
 * ```typescript
 * class ChatService {
 *   @TraceLLM({ modelName: 'gpt-4o', modelProvider: 'openai' })
 *   async chat(message: string) {
 *     return await this.client.chat.completions.create({...});
 *   }
 * }
 * ```
 */
export function TraceLLM(options: {
  modelName: string;
  modelProvider?: string;
  extractContent?: boolean;
}) {
  return function (
    _target: unknown,
    _propertyKey: string,
    descriptor: PropertyDescriptor
  ) {
    const originalMethod = descriptor.value;
    descriptor.value = traceLLM(originalMethod, options);
    return descriptor;
  };
}

/**
 * Decorator for tracing embedding calls on class methods.
 */
export function TraceEmbedding(options: { model: string }) {
  return function (
    _target: unknown,
    _propertyKey: string,
    descriptor: PropertyDescriptor
  ) {
    const originalMethod = descriptor.value;
    descriptor.value = traceEmbedding(originalMethod, options);
    return descriptor;
  };
}

/**
 * Decorator for tracing retrieval calls on class methods.
 */
export function TraceRetrieval(options: { vectorStore: string; embeddingModel?: string }) {
  return function (
    _target: unknown,
    _propertyKey: string,
    descriptor: PropertyDescriptor
  ) {
    const originalMethod = descriptor.value;
    descriptor.value = traceRetrieval(originalMethod, options);
    return descriptor;
  };
}

/**
 * Decorator for tracing tool calls on class methods.
 */
export function TraceTool(options: { toolName: string }) {
  return function (
    _target: unknown,
    _propertyKey: string,
    descriptor: PropertyDescriptor
  ) {
    const originalMethod = descriptor.value;
    descriptor.value = traceTool(originalMethod, options);
    return descriptor;
  };
}

/**
 * Decorator for tracing chain calls on class methods.
 */
export function TraceChain(options: { name: string }) {
  return function (
    _target: unknown,
    _propertyKey: string,
    descriptor: PropertyDescriptor
  ) {
    const originalMethod = descriptor.value;
    descriptor.value = traceChain(originalMethod, options);
    return descriptor;
  };
}

/**
 * Decorator for tracing agent calls on class methods.
 */
export function TraceAgent(options: { agentName: string; agentType?: string }) {
  return function (
    _target: unknown,
    _propertyKey: string,
    descriptor: PropertyDescriptor
  ) {
    const originalMethod = descriptor.value;
    descriptor.value = traceAgent(originalMethod, options);
    return descriptor;
  };
}
